### PR TITLE
Normalize reload method across source types

### PR DESCRIPTION
**Pull request by scothis**
- added a no-op reload method to video and raster sources
- guard against a layers without sources (background layers)
- fire 'change' event to trigger updates for visibility changes on raster,
  video and background layers

Issue: #1198

### CODE REVIEW

File: js/style/style.js [hunk 1]
```diff
@@ -370,7 +370,10 @@ Style.prototype = util.inherit(Evented, {
         layer = this.getReferentLayer(layer);
         layer.setLayoutProperty(name, value);
         this._broadcastLayers();
-        this.sources[layer.source].reload();
+        if (layer.source) {
```
Comment (jfirebaugh) [inline js/style/style.js]: What prompted this change?

Comment (scothis) [inline js/style/style.js]: Background layers don't have a source

File: js/style/style.js [hunk 2]
```diff
@@ -370,7 +370,10 @@ Style.prototype = util.inherit(Evented, {
...
+            this.sources[layer.source].reload();
+        }
+        this.fire('change');
```
Comment (jfirebaugh) [inline js/style/style.js]: Let's not do this in this PR. I am not sure it makes sense, and if it does make sense it should be done for the other mutating methods such as `setPaintProperty`.

File: js/style/style.js (hunk 1 above)
Comment (jfirebaugh) [inline js/style/style.js]: Right, ok. Can you:

* Make the same change for `setPaintProperty`
* Add tests for `setPaintProperty` and `setLayoutProperty` on a background layer

Comment (jfirebaugh) [inline js/style/style.js]: Oh wait, `setPaintProperty` won't have this issue. `setFilter` does, but background layers can't have filters.

**Comment by scothis** [comment]
@jfirebaugh can I get your eyes on this again?

**Comment by jfirebaugh** [comment]
This is three independent changes mixed in two commits. I'd like to see it as:

* A commit that adds tests that fail prior to the addition of no-op reloads on video and raster sources, and the addition of those no-op reloads.
* A commit that adds a test that fails prior to the `if (layer.source)` check, and the addition of that check.
* A commit that adds `this.fire('change');` to all relevant places, plus tests for that. (I'm not 100% sure we want to add this, it requires investigation of the current expectations of listeners to the change signal.)

**Comment by scothis** [comment]
@jfirebaugh thanks, I'll restructure

**Comment by jfirebaugh** [comment]
Here's what to think about with regards to firing the change event:

* Semantically, it makes sense -- changing a layout property is a style change.
* If we fire it for changing a layout property, we should fire it for changing a paint property, changing a filter, adding a layer, etc -- all the mutating methods.
* The [map listens for the change event](https://github.com/mapbox/mapbox-gl-js/blob/48e91df3388b31571ef326d96720549d1a11ba83/js/ui/map.js#L673-676) and [triggers style recalculation](https://github.com/mapbox/mapbox-gl-js/blob/master/js/ui/map.js#L594-597) in response to it.
* The public API for style changes is via the map -- `map.setPaintProperty`, etc. These methods also trigger recalculation in some situations.
* So if we add change event firing to the style methods, we may end up doing recalculation work twice.

**Comment by scothis** [comment]
The situation that caused me to fire the `change` event was toggling visibility of raster and background layers. For vector layers, reloading the source is enough to trigger a render. For raster and background layers, their is no immediate effect until another event causes a render.

Firing a `change` event multiple times from the same JS turn appears to be safe, as it sets a dirty flag and schedules the heavy work for a future turn. The actual work will only happen once as the dirty flag will be unset by the work.

I'll fire the `change` event for `setFilter()`, `setLayerProperty()` and `setPaintProperty()`. add/remove source/layer already fire more specialized events

**Comment by scothis** [comment]
@jfirebaugh updated into three discrete commits per your comments.

**Comment by jfirebaugh** [comment]
Thanks @scothis, this looks great!

**Pull request merged**